packages/vite/src/node/server/middlewares/indexHtml.ts

```diff
@@ -148,15 +149,23 @@ const processNodeUrl = (
         originalUrl !== '/' &&
         htmlPath === '/index.html')
     ) {
-      const devBase = config.base
-      const fullUrl = path.posix.join(devBase, url)
-      if (server && shouldPreTransform(url, config)) {
-        preTransformRequest(server, fullUrl, devBase)
-      }
-      return fullUrl
-    } else {
-      return url
+      url = path.posix.join(config.base, url)
+
+      preTransformUrl = url
+    } else if (url[0] === '/') {
+      preTransformUrl = url
+    } else if (url[0] === '.') {
+      preTransformUrl = path.posix.join(
+        config.base,
+        path.posix.dirname(htmlPath),
+        url,
+      )
+    }
+
+    if (preTransformUrl && server && shouldPreTransform(url, config)) {
```

Comment:
We should wrap all this part in a `if (server && shouldPreTransform(url, config))`

Reply:
To avoid the `path.posix.dirname`?

Reply:
pushed a refactor (and a fix) here https://github.com/vitejs/vite/pull/15303/commits/ba03a838f9be5fd62c7219441c557b30a9a03e69

Reply:
Yeah and also allow to isolate the code needed for the pre-transform part, thanks 👍